src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4567,5 +4419,452 @@ public int Current
+        [WorkItem(32316, "https://github.com/dotnet/roslyn/issues/32316")]
+        public void PatternBasedDisposal_TwoOverloads()
+        {
+            string source = @"
+using System.Threading.Tasks;
+class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+        }
+        System.Console.Write(""Done"");
+    }
+    public Enumerator GetAsyncEnumerator()
+    {
+        return new Enumerator();
+    }
+    public sealed class Enumerator
+    {
+        public async System.Threading.Tasks.Task<bool> MoveNextAsync()
+        {
+            System.Console.Write(""MoveNextAsync "");
+            await System.Threading.Tasks.Task.Yield();
+            return false;
+        }
+        public int Current
+        {
+            get => throw null;
+        }
+        public async System.Threading.Tasks.ValueTask DisposeAsync(int i = 0)
+        {
+            System.Console.Write(""DisposeAsync "");
+            await System.Threading.Tasks.Task.Yield();
+        }
+        public System.Threading.Tasks.ValueTask DisposeAsync(params string[] s)
+            => throw null;
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_IAsyncEnumerable }, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "MoveNextAsync DisposeAsync Done");
+        }
+
+        [ConditionalFact(typeof(WindowsDesktopOnly))]
+        [WorkItem(32316, "https://github.com/dotnet/roslyn/issues/32316")]
+        public void PatternBasedDisposal_NoExtensions()
+        {
+            string source = @"
+using System.Threading.Tasks;
+public class C
+{
+    public static async Task Main()
+    {
+        await foreach (var i in new C())
+        {
+        }
+        System.Console.Write(""Done"");
+    }
+    public Enumerator GetAsyncEnumerator()
+    {
+        return new Enumerator();
+    }
+    public sealed class Enumerator
+    {
+        public async System.Threading.Tasks.Task<bool> MoveNextAsync()
+        {
+            System.Console.Write(""MoveNextAsync "");
+            await System.Threading.Tasks.Task.Yield();
+            return false;
+        }
+        public int Current
+        {
+            get => throw null;
+        }
+    }
+}
+public static class Extension
+{
+    public static System.Threading.Tasks.ValueTask DisposeAsync(this C.Enumerator e) => throw null;
+}";
+            // extension methods do not contribute to pattern-based disposal
+            var comp = CreateCompilationWithTasksExtensions(new[] { source, s_IAsyncEnumerable }, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "MoveNextAsync Done");
+        }
+
+        [ConditionalFact(typeof(WindowsDesktopOnly))]
+        [WorkItem(32316, "https://github.com/dotnet/roslyn/issues/32316")]
+        public void PatternBasedDisposal_NoExtensions_TwoExtensions()
```

Comment:
>NoExtensions_TwoExtensions [](start = 41, length = 26)

Is this test necessary? Do we look up extension methods for the collection type?

Reply:
I think it is useful. See the tracking issue https://github.com/dotnet/roslyn/issues/32767
In this PR, I reject a binding if it is an extension, but we should really do the binding without considering extensions in the first place. So I added tests to show whether this has an impact or not. It does have an impact on `using` (we'll produce an ambiguity diagnostic), but does not impact `foreach` (shown here).

Reply:
The extension methods take the collection type not the enumerator type though, and would not be considered even if extension methods were supported for pattern-based dispose.

---
In reply to: [251030838](https://github.com/dotnet/roslyn/pull/32731#discussion_r251030838) [](ancestors = 251030838)

Reply:
Oops, sorry, same testing bug as above. Should be extensions on the enumerator